### PR TITLE
Python bindings for traffic

### PR DESCRIPTION
**Pull request by syershov**

### CODE REVIEW

File: xcode/omim.xcworkspace/contents.xcworkspacedata [hunk 1]
```diff
@@ -289,6 +289,9 @@
       location = "container:storage/storage.xcodeproj">
    </FileRef>
    <FileRef
+      location = "group:traffic/traffic.xcodeproj">
```
Comment (igrechuhin) [inline xcode/omim.xcworkspace/contents.xcworkspacedata]: container:

File: xcode/traffic/traffic.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+		67BECB5D1DDA44FD00FC4E99 /* traffic_info.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = traffic_info.cpp; sourceTree = "<group>"; };
+		67BECB5E1DDA44FD00FC4E99 /* traffic_info.hpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.h; path = traffic_info.hpp; sourceTree = "<group>"; };
+		67BECB671DDA466800FC4E99 /* traffic_tests.app */ = {isa = PBXFileReference; explicitFileType = wrapper.application; includeInIndex = 0; path = traffic_tests.app; sourceTree = BUILT_PRODUCTS_DIR; };
+		67BECB7F1DDA46A100FC4E99 /* traffic_info_test.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; path = traffic_info_test.cpp; sourceTree = "<group>"; };
+		67BECB841DDA474400FC4E99 /* libbase.a */ = {isa = PBXFileReference; lastKnownFileType = archive.ar; name = libbase.a; path = "/Users/darkserj/mapsme/omim/xcode/base/../../../omim-build/xcode/Debug/libbase.a"; sourceTree = "<absolute>"; };
```
Comment (igrechuhin) [inline xcode/traffic/traffic.xcodeproj/project.pbxproj]: path

File: xcode/traffic/traffic.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+		67BECB861DDA475500FC4E99 /* libplatform.a */ = {isa = PBXFileReference; lastKnownFileType = archive.ar; name = libplatform.a; path = "/Users/darkserj/mapsme/omim/xcode/platform/../../../omim-build/xcode/Debug/libplatform.a"; sourceTree = "<absolute>"; };
```
Comment (igrechuhin) [inline xcode/traffic/traffic.xcodeproj/project.pbxproj]: path

File: xcode/traffic/traffic.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+		67BECB881DDA476A00FC4E99 /* libcoding.a */ = {isa = PBXFileReference; lastKnownFileType = archive.ar; name = libcoding.a; path = "/Users/darkserj/mapsme/omim/xcode/coding/../../../omim-build/xcode/Debug/libcoding.a"; sourceTree = "<absolute>"; };
```
Comment (igrechuhin) [inline xcode/traffic/traffic.xcodeproj/project.pbxproj]: path

File: xcode/traffic/traffic.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+		67BECB8A1DDA478800FC4E99 /* testingmain.cpp */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.cpp.cpp; name = testingmain.cpp; path = /Users/darkserj/mapsme/omim/search/search_tests/../../testing/testingmain.cpp; sourceTree = "<absolute>"; };
```
Comment (igrechuhin) [inline xcode/traffic/traffic.xcodeproj/project.pbxproj]: path

File: xcode/traffic/traffic.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+		67BECB8D1DDA47B200FC4E99 /* libtomcrypt.a */ = {isa = PBXFileReference; lastKnownFileType = archive.ar; name = libtomcrypt.a; path = "/Users/darkserj/mapsme/omim/xcode/tomcrypt/../../../omim-build/xcode/Debug/libtomcrypt.a"; sourceTree = "<absolute>"; };
```
Comment (igrechuhin) [inline xcode/traffic/traffic.xcodeproj/project.pbxproj]: path

File: traffic/CMakeLists.txt [hunk 7]
```diff
@@ -10,4 +10,5 @@ set(
 
 add_library(${PROJECT_NAME} ${SRC})
 
-add_subdirectory(traffic_tests)
+add_subdirectory(pytraffic)
+#add_subdirectory(traffic_tests)
```
Comment (ygorshenin) [inline traffic/CMakeLists.txt]: Why? This line was uncommented.

File: traffic/pytraffic/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,45 @@
+project(pytraffic)
+
+check_pybindings()
+
+set(
+  SRC
+  bindings.cpp
+)
+add_compile_options(
```
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: Either move this to the root CMakeList or remove.

File: traffic/pytraffic/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    "-Wno-unused-local-typedef"
+)
+
+set(Boost_USE_MULTITHREADED ON)
+
+if (PLATFORM_MAC)
```
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: Why you missed linux? What's the difference?

File: traffic/pytraffic/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    ${PROJECT_NAME}
+    ${Qt5Widgets_LIBRARIES}
+    "-framework Cocoa"
+    "-framework IOKit"
+    "-framework QuartzCore"
+    "-framework SystemConfiguration"
+  )
+endif()
+
+if (PLATFORM_WIN OR PLATFORM_LINUX)
```
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: So, why not to:
```cmake
omim_link_libraries(${PROJECT_NAME} ${Qt5 stuff})
if (PLATFORM_MAC)
  ...
endif()
```

File: traffic/CMakeLists.txt (hunk 7 above)
Comment (syershov) [inline traffic/CMakeLists.txt]: Because @mpimenov didn't create CMakeList.txt for tests.

File: traffic/pytraffic/CMakeLists.txt (hunk 9 above)
Comment (syershov) [inline traffic/pytraffic/CMakeLists.txt]: Because on Linux we don't have static boost libs.

File: traffic/pytraffic/CMakeLists.txt (hunk 10 above)
Comment (syershov) [inline traffic/pytraffic/CMakeLists.txt]: This code will move to upper level. Here will be a something like use_platform().

File: traffic/CMakeLists.txt (hunk 7 above)
Comment (ygorshenin) [inline traffic/CMakeLists.txt]: Am I right that CMakeBuild was broken due to that missing CMakeLists?

File: traffic/pytraffic/CMakeLists.txt (hunk 8 above)
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: Please, leave a comment here.

File: traffic/pytraffic/CMakeLists.txt (hunk 9 above)
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: OK

File: traffic/pytraffic/CMakeLists.txt (hunk 10 above)
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: OK

File: traffic/CMakeLists.txt (hunk 7 above)
Comment (syershov) [inline traffic/CMakeLists.txt]: It's not missing, just hasn't been done yet.

File: traffic/pytraffic/CMakeLists.txt (hunk 8 above)
Comment (syershov) [inline traffic/pytraffic/CMakeLists.txt]: done

File: traffic/pytraffic/bindings.cpp [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+traffic::TrafficInfo::Coloring Deserialize(vector<uint8_t> const & data)
+{
+  traffic::TrafficInfo::Coloring coloring;
+  traffic::TrafficInfo::DeserializeTrafficData(data, coloring);
+  return coloring;
+}
+
+string RoadSegmentIdRepr(traffic::TrafficInfo::RoadSegmentId const & v)
```
Comment (syershov) [inline traffic/pytraffic/bindings.cpp]: It's wrapper for __repr__ method, not __str__.

File: traffic/pytraffic/bindings.cpp [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+#include <boost/python.hpp>
+#include <boost/python/suite/indexing/map_indexing_suite.hpp>
+
+namespace
+{
+using namespace boost::python;
+
+// Converts a vector<uint8_t> to/from Python str.
+struct vector_uint8t_to_str
```
Comment (syershov) [inline traffic/pytraffic/bindings.cpp]: done

File: traffic/pytraffic/bindings.cpp [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+BOOST_PYTHON_MODULE(pytraffic)
+{
+  using namespace boost::python;
+
+  // Register the to-python converters.
+  to_python_converter<vector<uint8_t>, vector_uint8t_to_str>();
+  vector_uint8t_from_python_str();
+
+  class_<traffic::TrafficInfo::RoadSegmentId>("RoadSegmentId", init<uint32_t, uint16_t, uint8_t>())
+    .def( "__repr__", &RoadSegmentIdRepr )
```
Comment (syershov) [inline traffic/pytraffic/bindings.cpp]: done

**Comment by syershov** [comment]
@ygorshenin PTAL

File: pyhelpers/pair.hpp [hunk 14]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include "std/utility.hpp"
+
+#include <boost/python.hpp>
+#include <boost/python/suite/indexing/vector_indexing_suite.hpp>
+
+namespace
```
Comment (ygorshenin) [inline pyhelpers/pair.hpp]: namespace pyhelpers?

File: pyhelpers/vector_uint8.hpp [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    const char * value = PyString_AsString(obj_ptr);
+    if (value == nullptr)
+      throw_error_already_set();
+    void * storage =
+        ((converter::rvalue_from_python_storage<vector<uint8_t>> *)data)->storage.bytes;
+    new (storage) vector<uint8_t>(value, value + PyString_Size(obj_ptr));
+    data->convertible = storage;
+  }
+};
+}  // namespace
```
Comment (ygorshenin) [inline pyhelpers/vector_uint8.hpp]: Add a blank line here.

File: traffic/pytraffic/CMakeLists.txt [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+)
+
+# Suppress boost-python warnings
+add_compile_options(
+    "-Wno-unused-local-typedef"
+)
+
+set(Boost_USE_MULTITHREADED ON)
+
+# For macOS we can use static linking, on Linux can't.
```
Comment (ygorshenin) [inline traffic/pytraffic/CMakeLists.txt]: on Linux we can't.

File: pyhelpers/vector_uint8.hpp (hunk 15 above)
Comment (syershov) [inline pyhelpers/vector_uint8.hpp]: done

File: pyhelpers/pair.hpp (hunk 14 above)
Comment (syershov) [inline pyhelpers/pair.hpp]: I think no.

**Comment by ygorshenin** [comment]
LGTM

**Pull request merged**